specs/FrameProcesssInfo.md

```diff
@@ -0,0 +1,513 @@
+                        continue;
+                    }
+                    // Traverse parent frame until find the parent main frame.
+                    while (parentFrameInfo.ParentFrameInfo != null)
+                    {
+                        firstLevelFrameInfo = parentFrameInfo;
+                        parentFrameInfo = parentFrameInfo.ParentFrameInfo;
+                    }
+                    renderProcessInfoStr += $" | frame Kind: other frame\n";
+                    renderProcessInfoStr += $" | parent main frame ID:{parentFrameInfo.FrameId}\n";
+                    renderProcessInfoStr += $" | parent first level frame ID:{firstLevelFrameInfo.FrameId}}}\n";
+                }
+                // </AssociatedFrameInfos>
+                string frameInfoRes = $"{frameInfoCount} frame info(s) found in renderer process ID: {processId}\n";
+                frameInfoRes += renderProcessInfoStr;
+                result += $"{frameInfoRes} \n";
+                processWithFrameCount++;
+            }
+            else 
+            {
+                otherProcessInfoStr += $"Process ID: {processId} | Process Kind: {kind}\n";
+            }
+
+        }
+        // </GetProcessInfosWithDetailsAsync>
+        string processCountStr = $"{processWithFrameCount} renderer process(es)\n\n";
+        MessageBox.Show(this, processCountStr + result + otherProcessInfoStr, "Process Frame Info List");
+    }
+    catch (NotImplementedException)
+    {
+        // If the runtime support is not there we probably want this
+        // to be a no-op.
+    }
+}
+```
+
+# API Details
+## C++
+```c++
+interface ICoreWebView2_18;
+interface ICoreWebView2Frame5;
+interface ICoreWebView2FrameInfo2;
+interface ICoreWebView2Environment14;
+interface ICoreWebView2ProcessInfo2;
+interface ICoreWebView2GetProcessInfosWithDetailsCompletedHandler;
+
+/// Receives the result of the `GetProcessInfosWithDetails` method.
+/// The result is written to the collection of `ProcessInfo`s provided
+/// in the `GetProcessInfosWithDetails` method call.
+[uuid(8e7d154c-e2ca-11ed-b5ea-0242ac120002), object, pointer_default(unique)]
+interface ICoreWebView2GetProcessInfosWithDetailsCompletedHandler : IUnknown {
+  HRESULT Invoke([in] HRESULT errorCode, [in] ICoreWebView2ProcessInfoCollection* value);
+}
+
+/// A continuation of the ICoreWebView2ProcessInfo interface.
+[uuid(982ae768-e2ca-11ed-b5ea-0242ac120002), object, pointer_default(unique)]
+interface ICoreWebView2ProcessInfo2 : ICoreWebView2ProcessInfo {
+  /// Gets the collection of the assocated `FrameInfo`s which are running in that
+  /// renderer process.
+  /// This is only available when process `Kind` is `COREWEBVIEW2_PROCESS_KIND_RENDERER`.
+  /// `frames` is `null` for all other process kinds.
+  /// Note that `frames` can be `null` if there's no frame running in that process.
+  ///
+  /// \snippet ProcessComponent.cpp AssociatedFrameInfos
+  [propget] HRESULT AssociatedFrameInfos(

```

Comment:
Do we need to add that it will be null if it is called not from ICoreWebView2GetProcessInfosWithDetailsCompletedHandler? The same question about ICoreWebView2FrameInfo2 ::ParentFrameInfo. Do not remember about ICoreWebView2FrameInfo2::FrameId.
Or we need to return error in those cases?

Reply:
It's good to return error in those cases. Added.

Reply:
Updated the document. To keep consistency, we decide to return empty/null instead of return error code for those cases.

Reply:
Then `Assocated `FrameInfo`s is empty if there's no frame running in that renderer process.` is not true - it can be also that method is not called from handler. Do we need to distinguish between those two cases?

Reply:
Yeah, there're 3 cases which returns empty frameInfo collection
1. Renderer process has no actively running frames. 
2. From other process Kind. I think can combine with 1) which say no active running frames on that __process__.l
3. Not called from handlers. 

We decide to return empty list for all those cases. Need a rephase suggestion @david-risney

Reply:
Its better to reference the async method `GetProcessInfosWithDetails` rather than the delegate interface `ICoreWebView2GetProcessInfosWithDetailsCompletedHandler` which is mostly an implementation detail as far as the end developer is concerned and is mostly irrelevant esp for WinRT and .NET.

It is useful to separate out the cases where it is empty because the end dev did something wrong (not a renderer process & didn't call GetProcessInfosWithDetails - end dev made a mistake in their code) versus the collection is empty because there just aren't any (valid case the end dev needs to consider). Also we can state it in the positive first (use GetProcessInfosWithDetails and only available for renderer processes) and optionally follow that with explicit statements about the negative case (GetProcessInfos will not populate this property, and this property is always empty for non-renderer processes). Plus an explicit note that in some cases a renderer process may not have any active frames and so the AssociatedFrameInfos may still be empty.

```
  /// The collection of associated `FrameInfo`s which are actively running
  /// (showing UI elements) in this renderer process. `AssociatedFrameInfos` will only
  /// be populated when obtained via calling `CoreWebView2.GetProcessInfosWithDetails` and
  /// when this `CoreWebView2ProcessInfo` corresponds to a renderer process.
  /// `CoreWebView2ProcessInfo` objects obtained via `CoreWebView2.GetProcessInfos` or for non-renderer
  /// processes will always have an empty `AssociatedFrameInfos`. The `AssociatedFrameInfos` may also be
  /// be empty for renderer processes that have no active frames.
  /// \snippet ProcessComponent.cpp AssociatedFrameInfos
```